File: src/scan.ts
import {
  MAX_PAYMENT_SIZE,
  initNode,
  makeRandomHash,
} from './util';

// myChannelId is the channel ID we have to connect us to the scanChannelId
// scanChannelId is the channel ID of the channel we want to know the balance of
export async function scanChannel(
  myChannelId: string,
  scanChannelId: string,
) {
  // Grab some initial data and make sure everything came back OK
  const node = await initNode();
  const myChannels = (await node.listChannels()).channels;
  const myChannel = myChannels.find(c => c.chanId === myChannelId);
  const channel = await node.getChanInfo({ chanId: scanChannelId });

  if (!myChannel) {
    throw new Error(`You have no channel with ID '${myChannelId}'`);
  }
  if (!channel) {
    throw new Error(`Unknown channel with ID '${scanChannelId}'`);
  }

  // Determine the max we can scan, and who's the "receiver"
  const max = Math.min(
    parseInt(channel.capacity, 10),
    parseInt(myChannel.localBalance, 10),
    MAX_PAYMENT_SIZE,
  );
  const destPubkey = myChannel.remotePubkey === channel.node1Pub
    ? channel.node2Pub
    : channel.node1Pub;

  // Loop send bogus payments until we find the balance
  let low = 0;
  let high = max;

  console.log(`Beginning scan of channel ${scanChannelId}, max scannable ${max} satoshis...`);
  while (high - low > 1) {
    const testAmount = Math.ceil((low + high) / 2);
    console.log(`Probing with a ${testAmount} sat payment...`);

    const res = await node.sendPaymentSync({
      destString: destPubkey,
      amt: testAmount.toString(),
      paymentHashString: makeRandomHash(),
      outgoingChanId: myChannel.chanId,
      finalCltvDelta: 144,
    });
    const err = res.paymentError;

    // Depending on the error, raise or lower the amount. The route note having
    // enough capacity comes in many shapes and sizes of error, so we have to
    // check for a few types here.
    if (err.includes('UnknownPaymentHash')) {
      console.log('Amount was too low, raising lower bound...');
      low = testAmount;
    } else if (
      err.includes('unable to find a path') ||
      err.includes('insufficient') ||
      err.includes('TemporaryChannelFailure')
    ) {
      console.log('Amount was too high, lowering upper bound...');
      high = testAmount;
    } else {
      throw new Error(`Unknown error occured when trying to scan: ${err}`);
    }
  }

  console.log(`Balance for channel ${scanChannelId} is between ${low} and ${high} satoshis!`);
  if (high === max) {
    console.log(`NOTE: The balance exceeds the height we were able to scan, so it may be larger than ${max} satoshis`);
  }
}

File: python/scan/scan.py
import math
import rpc_pb2 as ln
from scan.util import MAX_PAYMENT_SIZE, init_node, make_random_hash

# my_channel_id is the channel ID we have to connect us to the scanChannelId
# scan_channel_id is the channel ID of the channel we want to know the balance of
def scan_channel(my_channel_id: int, scan_channel_id: int):
    # Grab some initial data and make sure everything came back OK
    node = init_node()
    my_channels = node.ListChannels(ln.ListChannelsRequest()).channels
    my_channel = [c for c in my_channels if c.chan_id == my_channel_id][0]
    channel = node.GetChanInfo(ln.ChanInfoRequest(chan_id=scan_channel_id))

    if not my_channel:
        raise Exception(f'You have no channel with ID "{my_channel_id}"')
    if not channel:
        raise Exception(f'Unknown channel with ID "{scan_channel_id}"')

    # Determine the max we can scan, and who's the "receiver"
    maximum = min(
        int(channel.capacity),
        int(my_channel.local_balance),
        MAX_PAYMENT_SIZE,
    )
    dest_pubkey = channel.node2_pub if my_channel.remote_pubkey == channel.node1_pub else channel.node2_pub

    # Loop send bogus payments until we find the balance
    low = 0
    high = maximum

    print(f'Beginning scan of channel {scan_channel_id}, max scannable {maximum} satoshis...')
    while high - low > 1:
        test_amount = math.ceil((low + high) / 2)
        print(f'Probing with a {test_amount} sat payment...')

        res = node.SendPaymentSync(ln.SendRequest(
            dest_string=dest_pubkey,
            amt=test_amount,
            payment_hash_string=make_random_hash(),
            outgoing_chan_id=my_channel.chan_id,
            final_cltv_delta=144,
        ))
        err = res.payment_error

        # Depending on the error, raise or lower the amount. The route note having
        # enough capacity comes in many shapes and sizes of error, so we have to
        # check for a few types here.
        if 'UnknownPaymentHash' in err:
            print('Amount was too low, raising lower bound...')
            low = test_amount
        elif 'unable to find a path' in err or \
                'insufficient' in err or \
                'TemporaryChannelFailure' in err:
            print('Amount was too high, lowering upper bound...')
            high = test_amount
        else:
            raise Exception(f'Unknown error occured when trying to scan: {err}')

    print(f'Balance for channel ${scan_channel_id} is between {low} and {high} satoshis!')
    if high == maximum:
        print(f'NOTE: The balance exceeds the height we were able to scan, so it may be larger than {maximum} satoshis')

